src/infra/core/src/compact_service_impl.rs

```diff
@@ -0,0 +1,417 @@
+                    .int_err()?
+                    .sort(vec![col(offset_column).sort(true, false)])
+                    .int_err()?;
+
+                let new_file_path = compact_dir_path.join(format!("merge-slice-{index}").as_str());
+
+                data_frame
+                    .write_parquet(
+                        new_file_path.to_str().unwrap(),
+                        datafusion::dataframe::DataFrameWriteOptions::new()
+                            .with_single_file_output(true),
+                        None,
+                    )
+                    .await
+                    .int_err()?;
+                data_slice_batch_info.new_file_path = Some(new_file_path);
+            }
+        }
+
+        Ok(())
+    }
+
+    fn create_run_compact_dir(&self) -> Result<PathBuf, CompactError> {
+        let compact_dir_path = self
+            .run_info_dir
+            .join(get_random_name(Some("compact-"), 10));
+        fs::create_dir_all(&compact_dir_path).int_err()?;
+        Ok(compact_dir_path)
+    }
+
+    async fn commit_new_blocks(
+        &self,
+        dataset: Arc<dyn Dataset>,
+        chain_files_info: &ChainFilesInfo,
+    ) -> Result<(Vec<Url>, Multihash), CompactError> {
+        let chain = dataset.as_metadata_chain();
+        let mut current_head = chain_files_info.old_head.clone();
+        let mut old_data_slices: Vec<Url> = vec![];
+
+        for data_slice_batch in chain_files_info.data_slice_batches.iter().rev() {
+            match data_slice_batch {
+                DataSliceBatch::SingleBlock(block_hash) => {
+                    let block = chain.get_block(block_hash).await.int_err()?;
+
+                    let commit_result = dataset
+                        .commit_event(
+                            block.event,
+                            CommitOpts {
+                                block_ref: &BlockRef::Head,
+                                system_time: Some(self.time_source.now()),
+                                prev_block_hash: Some(Some(&current_head)),
+                                check_object_refs: false,
+                                update_block_ref: false,
+                            },
+                        )
+                        .await
+                        .int_err()?;
+                    current_head = commit_result.new_head;
+                }
+                DataSliceBatch::CompactedBatch(data_slice_batch_info) => {
+                    let add_data_file =
+                        OwnedFile::new(data_slice_batch_info.new_file_path.as_ref().unwrap());
+                    let new_offset_interval = OffsetInterval {
+                        start: data_slice_batch_info.lower_bound.start_offset,
+                        end: data_slice_batch_info.upper_bound.end_offset,
+                    };
+
+                    let (new_data, _) = dataset
+                        .prepare_objects(Some(new_offset_interval), Some(&add_data_file), None)
```

Comment:
Why do you need this low-level access to `prepare/commit_objects`? Why can't we use the high-level `commit_add_data` function instead?

I would really like to avoid exposing such low-level functions in `Dataset` interface so would like to understand why this was necessary.

To me it looked like the compacted dataset could be constructed using same high-level `commit_add_data` / `commit_event` primitives and the only thing we were missing was the ability to skip updating the `head` ref until the whole process is done.

Reply:
Refactored